rfcs/20220307-monitoring_across_the_stack.md

```diff
@@ -0,0 +1,353 @@
+
+## Purpose
+
+This RFC proposes we introduce [Prometheus](https://prometheus.io) and [Grafana](https://grafana.com/) to monitor our infrastructure. It also proposes that we use our existing Postgres instance for storing the gathered metrics.
+
+Prometheus is a time series database with an advanced query language and the ability to configure aggregate data transformations (for example, based on a given metric you can also tell Prometheus to calculate a running mean of the metric from over the last 5 minutes, or any other span of time). It includes anomaly detection as well as integration with many tools you can use for alerting like Slack, PagerDuty, or email. There are [official and unofficial client libraries](https://prometheus.io/docs/instrumenting/clientlibs/) for all the langauges we currently use or could conceivably use in the future including Python, Node.js, Go, PHP, and JVM languages.
+
+Grafana is a data visualization tool that supports building complex dashboards. It has the ability to integrate with nearly any source of data you could imagine, including our existing catalog and API databases. In the future we could use it to monitor cataloging progress over time for example. It's a tool good for many things in addition to visualizing metrics.
+
+The Prometheus and Grafana services encompass the following key features that are important for maintaining the health and stability of our services:
+
+* Metrics gathering
+* Dashboard creation for making it quick and easy to understand the current state of our service health
+* Anomaly detection and alarming so that if things get weird, we'll know about it as soon as possible
+
+Additionally, both are free software. Prometheus is Apache 2 licensed and Grafana is AGPL-3.0. Both are actively developed in the open with heavy community involvement. They are also both widely deployed with lots of blog posts discussing configuration, use cases, and troubleshooting.
+
+## Self-hosted vs. hosted
+
+I will defer this decision to @rbadillap because he has way more knowledge about what will be affordable for us both from a raw cost perspective and from a time-spent-maintaining perspective. My gut says that in the interest of getting monitoring sooner than later we could pay for hosted Prometheus and Grafana to start with and eventually migrate to self-hosted. AWS has hosted offerrings for both. I believe we could use our existing Postgres for them as well. I'm assuming that AWS hasn't added anything that would lock us into their hosted version of these services, but I could be wrong about that.
```

Comment:
That's a good point.

I need to revise this section. Grafana by default uses SQLite (makes provisioning easy). I wonder if we can survive with just that for now.

I'd hate to spin up a second RDS instance just for monitoring, just from a cost perspective. @rbadillap do you have any thoughts here? Can we get away with using SQLite in production for Grafana or is that going to bite us eventually?

Reply:
My gut says that we should spin up a new instance and that it wouldn't cost _too_ much (if Grafana can survive on sqlite it can certainly survive on a small postgres instance).

Reply:
Yeah if we don't care about cost as suggested by @zackkrida then just use a new instance. It will be more resilient into the future anyway.